# ===========================================================
# 05 Reading Tabular Data Using Pandas

# Objectives
- Import the Pandas library.
- Use Pandas to load a CSV data set.
- Get summary information from a Pandas DataFrame.
- Download online data using Pandas.

## Pandas
- Pandas is a widely-used Python library for statistics and plotting
- Its focus is tabular data
- It is similar to R in that it uses a structure called a dataframes.
- Dataframes can contain multiple data types

![Data Frame ](https://pandas.pydata.org/pandas-docs/stable/_images/01_table_dataframe.svg)

Source: <https://pandas.pydata.org/pandas-docs/stable/_images/01_table_dataframe.svg>

- Pandas can read all kinds of tabular data

![Data Processed by Pandas](https://pandas.pydata.org/pandas-docs/stable/_images/02_io_readwrite.svg)

Source: <https://pandas.pydata.org/pandas-docs/stable/_images/02_io_readwrite.svg>

In [1]:
# 1. Run this cell to download the data
# 2. Open the downloaded files to get a sense of the data

# Downloads a zip file from Carpentries webpage with Gapminder data
!wget http://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip .
# Unzips the file
!unzip python-novice-gapminder-data.zip

--2020-06-11 15:52:06--  http://swcarpentry.github.io/python-novice-gapminder/files/python-novice-gapminder-data.zip
Resolving swcarpentry.github.io (swcarpentry.github.io)... 185.199.110.153, 185.199.111.153, 185.199.108.153, ...
Connecting to swcarpentry.github.io (swcarpentry.github.io)|185.199.110.153|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 38471 (38K) [application/zip]
Saving to: ‘python-novice-gapminder-data.zip’

python-novice-gapmi 100%[===================>]  37.57K   189KB/s    in 0.2s    

2020-06-11 15:52:07 (189 KB/s) - ‘python-novice-gapminder-data.zip’ saved [38471/38471]

--2020-06-11 15:52:07--  http://./
Resolving . (.)... failed: Temporary failure in name resolution.
wget: unable to resolve host address ‘.’
FINISHED --2020-06-11 15:52:07--
Total wall clock time: 0.4s
Downloaded: 1 files, 38K in 0.2s (189 KB/s)
Archive:  python-novice-gapminder-data.zip
  inflating: data/gapminder_all.csv  
  inflating: data/gapminder_gdp_africa.csv  
 

- Load Pandas with `import pandas as pd`

In [2]:
# Import the pandas library
import pandas as pd

# Use `read_csv` to read the gapminder data
data = pd.read_csv('data/gapminder_gdp_oceania.csv')
print(data)

       country  gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
0    Australia     10039.59564     10949.64959     12217.22686   
1  New Zealand     10556.57566     12247.39532     13175.67800   

   gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  \
0     14526.12465     16788.62948     18334.19751     19477.00928   
1     14463.91893     16046.03728     16233.71770     17632.41040   

   gdpPercap_1987  gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  \
0     21888.88903     23424.76683     26997.93657     30687.75473   
1     19007.19129     18363.32494     21050.41377     23189.80135   

   gdpPercap_2007  
0     34435.36744  
1     25185.00911  


- The columns in a dataframe are the observed variables, and the rows are the observations.
- Pandas uses backslash `\` to show wrapped lines when output is too wide to fit the screen.

## `index_col`
- Use `index_col` to specify that a column's values should be used as row identifiers.

In [3]:
data = pd.read_csv('data/gapminder_gdp_oceania.csv', index_col='country')
print(data)

             gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
country                                                                       
Australia       10039.59564     10949.64959     12217.22686     14526.12465   
New Zealand     10556.57566     12247.39532     13175.67800     14463.91893   

             gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                       
Australia       16788.62948     18334.19751     19477.00928     21888.88903   
New Zealand     16046.03728     16233.71770     17632.41040     19007.19129   

             gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                      
Australia       23424.76683     26997.93657     30687.75473     34435.36744  
New Zealand     18363.32494     21050.41377     23189.80135     25185.00911  


- Use `DataFrame.info` to find out more about a dataframe.

In [4]:
# `info()` is a method of data
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2 entries, Australia to New Zealand
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   gdpPercap_1952  2 non-null      float64
 1   gdpPercap_1957  2 non-null      float64
 2   gdpPercap_1962  2 non-null      float64
 3   gdpPercap_1967  2 non-null      float64
 4   gdpPercap_1972  2 non-null      float64
 5   gdpPercap_1977  2 non-null      float64
 6   gdpPercap_1982  2 non-null      float64
 7   gdpPercap_1987  2 non-null      float64
 8   gdpPercap_1992  2 non-null      float64
 9   gdpPercap_1997  2 non-null      float64
 10  gdpPercap_2002  2 non-null      float64
 11  gdpPercap_2007  2 non-null      float64
dtypes: float64(12)
memory usage: 208.0+ bytes


*   This is a `DataFrame`
*   Two rows named `'Australia'` and `'New Zealand'`
*   Twelve columns, each of which has two actual 64-bit floating point values.
*   Uses 208 bytes of memory.

## Attributes
- The `DataFrame.columns` attribute stores information about the dataframe's columns.
- Note that this is a varaible, *not* a method.
  - It doesn't have `()`
*   Called a *member variable*, just a *member*, or an *attribute*.

In [5]:
print(data.columns)

Index(['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967',
       'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987',
       'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007'],
      dtype='object')


- Use `DataFrame.T` to transpose a dataframe.

*   Sometimes want to treat columns as rows and vice versa.
*   Transpose doesn't copy the data, just changes the program's view of it.

In [6]:
print(data.T)

country           Australia  New Zealand
gdpPercap_1952  10039.59564  10556.57566
gdpPercap_1957  10949.64959  12247.39532
gdpPercap_1962  12217.22686  13175.67800
gdpPercap_1967  14526.12465  14463.91893
gdpPercap_1972  16788.62948  16046.03728
gdpPercap_1977  18334.19751  16233.71770
gdpPercap_1982  19477.00928  17632.41040
gdpPercap_1987  21888.88903  19007.19129
gdpPercap_1992  23424.76683  18363.32494
gdpPercap_1997  26997.93657  21050.41377
gdpPercap_2002  30687.75473  23189.80135
gdpPercap_2007  34435.36744  25185.00911


## Summary Statistics
- Use `DataFrame.describe()` to get summary statistics about data.

- DataFrame.describe() gets the summary statistics of only the columns that have numerical data. 
  All other columns are ignored.

In [7]:
# 1. Print the summary statistics for our dataframe
print(data.describe())

       gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
count        2.000000        2.000000        2.000000        2.000000   
mean     10298.085650    11598.522455    12696.452430    14495.021790   
std        365.560078      917.644806      677.727301       43.986086   
min      10039.595640    10949.649590    12217.226860    14463.918930   
25%      10168.840645    11274.086022    12456.839645    14479.470360   
50%      10298.085650    11598.522455    12696.452430    14495.021790   
75%      10427.330655    11922.958888    12936.065215    14510.573220   
max      10556.575660    12247.395320    13175.678000    14526.124650   

       gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
count         2.00000        2.000000        2.000000        2.000000   
mean      16417.33338    17283.957605    18554.709840    20448.040160   
std         525.09198     1485.263517     1304.328377     2037.668013   
min       16046.03728    16233.717700    17632.410

# Exercise
1. `read_csv()` can download data directly from a webpage.
   Download a dataset called the Titanic Data Set by passing
   the following URL to `read_csv()` instead of a file path.
   Put the new dataframe in a variable called `titanic`.
2. Use `titanic.head()` to have a look at the new dataframe.

**Data URL:**
<https://github.com/pandas-dev/pandas/raw/master/doc/data/titanic.csv>

In [8]:
# 1.
import pandas as pd
titanic = pd.read_csv('https://github.com/pandas-dev/pandas/raw/master/doc/data/titanic.csv')
# 2.
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Objectives
- Import the Pandas library.
- Use Pandas to load a CSV data set.
- Get summary information from a Pandas DataFrame.
- Download online data using Pandas.